### PR TITLE
Feat/dev74 gather the historical delegated supply of all supported da os

### PR DESCRIPTION
**Pull request by LeonardoVieira1630**
## Summary

**Aggregated Delegated Supply Endpoint** - A new GraphQL endpoint that aggregates delegation percentage metrics across all DAOs

---

### What it does
Provides a unified view of delegation percentage across all supported DAOs by calculating the mean delegation percentage for each day in a specified date range.

#### **Indexer API** (`apps/indexer`)
- **New endpoint**: `GET /delegation-percentage`
- **Features**:
  - Forward-fill algorithm for missing dates
  - Cursor-based pagination (`after`, `before`, `limit`)
  - Efficient query optimization using cursor parameters
  - Previous values lookup for accurate forward-fill across data gaps
  - Automatic adjustment of startDate when requested date is before available data

**Files:**
- `src/api/controller/delegation-percentage.controller.ts` - REST endpoint
- `src/api/services/delegation-percentage/delegation-percentage.service.ts` - Business logic (339 lines)
- `src/api/services/delegation-percentage/delegation-percentage.service.test.ts` - 19 unit tests
- `src/api/repositories/delegation-percentage.repository.ts` - Database queries
- `src/api/mappers/delegation-percentage.ts` - Zod schemas and type inference

#### **API Gateway** (`apps/api-gateway`)
- **New GraphQL query**: `aggregatedDelegatedSupply`
- **Features**:
  - Fetches from all DAO indexers in parallel
  - Aligns DAO responses to ensure all DAOs have data in the same date range
  - Aggregates using mean calculation
  - Handles pagination metadata correctly after alignment

**Files:**
- `src/resolvers/aggregated-delegated-supply.ts` - Main resolver (263 lines)
- `src/resolvers/aggregated-delegated-supply.test.ts` - 10 unit tests
- `meshrc.ts` - GraphQL schema definition with documentation

### Test Coverage
- **Indexer**: 19 unit tests covering:
  - Forward-fill with gaps
  - Previous values lookup
  - Edge cases (startDate before data, after-only queries, before-only queries)
  - Pagination
  - Error handling

- **Gateway**: 10 unit tests covering:
  - Mean percentage calculation
  - DAO alignment logic
  - Pagination response building
  - Edge cases (empty data, misaligned dates)

### CODE REVIEW

File: apps/indexer/src/api/repositories/delegation-percentage.repository.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      .from(daoMetricsDayBucket)
+      .where(and(...queryFilters.filter(Boolean)))
+      .orderBy(
+        orderDirection === "desc"
+          ? desc(daoMetricsDayBucket.date)
+          : asc(daoMetricsDayBucket.date),
+      );
+
+    const rows =
+      limit !== undefined ? await baseQuery.limit(limit) : await baseQuery;
```
Comment (PedroBinotto) [inline apps/indexer/src/api/repositories/delegation-percentage.repository.ts]: is there a cleaner way to do this?

File: apps/indexer/src/api/repositories/delegation-percentage.repository.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    return rows as DaoMetricRow[];
```
Comment (PedroBinotto) [inline apps/indexer/src/api/repositories/delegation-percentage.repository.ts]: it might be a good idea/ better separation of concerns if you return the database type natively from the repo function and map it to a custom interface later on, in the mapper layer

File: apps/indexer/src/api/repositories/delegation-percentage.repository.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      .where(
+        and(
+          lte(daoMetricsDayBucket.date, BigInt(beforeDate)),
+          inArray(daoMetricsDayBucket.metricType, [metricType]),
+        ),
+      )
+      .orderBy(desc(daoMetricsDayBucket.date))
+      .limit(1);
+
+    return rows[0] ? (rows[0] as DaoMetricRow) : null;
```
Comment (PedroBinotto) [inline apps/indexer/src/api/repositories/delegation-percentage.repository.ts]: same thing about the return type as the last comment

File: apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+
+export class DelegationPercentageService {
+  constructor(private readonly repository: DelegationPercentageRepository) {}
+
+  /**
+   * Main method to get delegation percentage data with forward-fill and pagination
+   */
+  async getDelegationPercentage(
+    filters: DelegationPercentageFilters,
+  ): Promise<DelegationPercentageResponse> {
```
Comment (PedroBinotto) [inline apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts]: I think it's better to decouple the service layer from the web/controller type namespace; maybe extract the type conversion to the mapper layer and either call it from within this class or in the controller after the data has been resolved

File: apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+      after,
+      before,
+      startDate,
+      endDate,
+      orderDirection = "asc",
+      limit = 100,
+    } = filters;
+
+    // 1. Get initial values for proper forward-fill
+    const referenceDate = startDate || after;
```
Comment (PedroBinotto) [inline apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts]: double check this pls

```suggestion
    const referenceDate =  after || startDate;
```

File: apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+    before?: string,
+  ): PageInfo {
+    return {
+      hasNextPage,
+      hasPreviousPage: !!after || !!before,
+      endCursor:
+        items.length > 0 ? (items[items.length - 1]?.date ?? null) : null,
+      startCursor: items.length > 0 ? (items[0]?.date ?? null) : null,
+    };
+  }
```
Comment (PedroBinotto) [inline apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts]: might be a better fit for a mapper method

**Review by PedroBinotto — CHANGES_REQUESTED** [review]

File: apps/indexer/src/api/controller/delegation-percentage.controller.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+      summary: "Get delegation percentage day buckets with forward-fill",
+      tags: ["metrics"],
+      request: {
+        query: z.object({
+          after: z.string().optional(),
+          before: z.string().optional(),
+          startDate: z.string().optional(),
+          endDate: z.string().optional(),
+          orderDirection: z.enum(["asc", "desc"]).default("asc"),
+          limit: z.coerce.number().int().positive().max(1000).default(100),
```
Comment (pikonha) [inline apps/indexer/src/api/controller/delegation-percentage.controller.ts]:
```suggestion
          limit: z.coerce.number().max(1000).default(365),
```

File: apps/indexer/src/api/mappers/delegation-percentage.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+export const DelegationPercentageItemSchema = z.object({
+  date: z.string(),
+  high: z.string(),
+});
+
+export const PageInfoSchema = z.object({
+  hasNextPage: z.boolean(),
+  hasPreviousPage: z.boolean(),
+  endCursor: z.string().nullable(),
+  startCursor: z.string().nullable(),
```
Comment (pikonha) [inline apps/indexer/src/api/mappers/delegation-percentage.ts]:
```suggestion
  endDate: z.string().nullable(),
  startDate: z.string().nullable(),
```

File: apps/indexer/src/api/repositories/delegation-percentage.repository.ts (hunk 2 above)
Comment (pikonha) [reply apps/indexer/src/api/repositories/delegation-percentage.repository.ts]: try using [this drizzle way of fetching db entities](https://github.com/blockful/anticapture/blob/main/apps/indexer/src/api/repositories/transactions.repository.ts#L11) which is already typed

File: apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+  async getDelegationPercentage(
+    filters: DelegationPercentageFilters,
+  ): Promise<DelegationPercentageResponse> {
+    const {
+      after,
+      before,
+      startDate,
+      endDate,
+      orderDirection = "asc",
+      limit = 100,
```
Comment (pikonha) [inline apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts]:
```suggestion
      orderDirection,
      limit,
```

File: apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+   * Adjusts startDate to the first real data date if requested startDate is before any data
+   * Returns the original startDate if it's within or after available data range
+   */
+  private adjustStartDateToFirstRealData(
+    startDate: string | undefined,
+    after: string | undefined,
+    dateMap: Map<string, DateData>,
+    initialValues: { delegated: bigint; total: bigint },
+  ): string | undefined {
+    const referenceDate = startDate || after;
```
Comment (pikonha) [inline apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts]:
```suggestion
    const referenceDate = after || startDate;
```

**Review by pikonha — COMMENTED** [review]
please improve the documentation to register all the mapped edge cases

File: apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+  DelegationPercentageRepository,
+  DaoMetricRow,
+} from "@/api/repositories/delegation-percentage.repository";
+import type {
+  DelegationPercentageItem,
+  PageInfo,
+  DelegationPercentageResponse,
+} from "@/api/mappers/delegation-percentage";
+
+export interface DelegationPercentageFilters {
```
[the remark this answers is not in the document]
Comment (PedroBinotto) [reply apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts]: ☝️

File: apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+    const lastDate = endDate
+      ? BigInt(endDate)
+      : datesFromDb[datesFromDb.length - 1];
+
+    if (!firstDate || !lastDate) {
+      return allDates;
+    }
+
+    // Generate all days in range (86400 seconds = 1 day)
+    const ONE_DAY = 86400n;
```
[the remark this answers is not in the document]
Comment (PedroBinotto) [reply apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts]: marked "resolved" with no address;

also, this could just be a reference to the `SECONDS_IN_DAY` const

File: apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts (hunk 5 above)
Comment (PedroBinotto) [reply apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts]: marked "resolved" with no address;

File: apps/api-gateway/src/resolvers/aggregated-delegated-supply.ts [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+ * Helper to get the first (oldest) date from items based on order direction
+ */
+function getFirstDate(
+  items: { date: string; high: string }[],
+  orderDirection?: string
+): bigint {
+  if (items.length === 0) throw new Error('No items to get first date from');
+  const firstItem =
+    orderDirection === 'desc' ? items[items.length - 1] : items[0];
+  return BigInt(firstItem.date);
```
[the remark this answers is not in the document]
Comment (PedroBinotto) [reply apps/api-gateway/src/resolvers/aggregated-delegated-supply.ts]: reiterating this concern; believe it remains relevant

File: apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts [hunk 14]
```diff
@@ -0,0 +1,386 @@
...
+      // Calculate percentage (avoid division by zero)
+      // Returns as string with 2 decimal places (e.g., "11.74" for 11.74%)
+      const percentage =
+        lastTotal > 0n ? Number((lastDelegated * 10000n) / lastTotal) / 100 : 0;
+
+      return {
+        date: dateStr,
+        high: percentage.toFixed(2),
+      };
+    });
```
Comment (PedroBinotto) [inline apps/indexer/src/api/services/delegation-percentage/delegation-percentage.service.ts]: Could something like this work? looks a bit more declarative/stateless-ish (i hate variables with a burning passion)

```suggestion
    const process = (
      dates: bigint[],
      delegated: bigint,
      total: bigint,
    ): DelegationPercentageItem[] => {
      if (dates.length === 0) return [];
      
      const [date, ...rest] = dates;
      const data = dateMap.get(date.toString());
      const newDelegated = data?.delegated ?? delegated;
      const newTotal = data?.total ?? total;
      const percentage = newTotal > 0n ? Number((newDelegated * 10000n) / newTotal) / 100 : 0;
      
      return [
        { date: date.toString(), high: percentage.toFixed(2) },
        ...process(rest, newDelegated, newTotal),
      ];
    };
    
    return process(allDates, initialValues.delegated, initialValues.total);
```

File: apps/indexer/src/api/repositories/delegation-percentage.repository.ts [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    const rows = await db.query.daoMetricsDayBucket.findMany({
+      where: and(
+        lte(daoMetricsDayBucket.date, BigInt(beforeDate)),
+        inArray(daoMetricsDayBucket.metricType, [metricType]),
+      ),
+      orderBy: desc(daoMetricsDayBucket.date),
+      limit: 1,
+    });
+
+    return rows[0] ?? null;
```
Comment (pikonha) [inline apps/indexer/src/api/repositories/delegation-percentage.repository.ts]:
```suggestion
        return await db.query.daoMetricsDayBucket.findFirst({
      where: and(
        lte(daoMetricsDayBucket.date, BigInt(beforeDate)),
        inArray(daoMetricsDayBucket.metricType, [metricType]),
      ),
      orderBy: desc(daoMetricsDayBucket.date),
    });
```
